src/Renderer/ThreeExtended/FirstPersonControls.js

```diff
@@ -0,0 +1,129 @@
+    40: { method: 'translateZ', sign: 1 }, // BACKWARD: down key
+    37: { method: 'translateX', sign: -1 }, // STRAFE_LEFT: left key
+    39: { method: 'translateX', sign: 1 }, // STRAFE_RIGHT: right key
+    33: { method: 'translateY', sign: 1 }, // UP: PageUp key
+    34: { method: 'translateY', sign: -1 }, // DOWN: PageDown key
+};
+
+
+class FirstPersonControls extends THREE.EventDispatcher {
+    constructor(view, options = {}) {
+        super();
+        this.camera = view.camera.camera3D;
+        this.view = view;
+        this.moves = new Set();
+        this.moveSpeed = options.moveSpeed || 10; // backward or forward move speed in m/s
+        this._isUserInteracting = false;
+        this._onMouseDownMouseX = 0;
+        this._onMouseDownMouseY = 0;
+        this._onMouseDownLon = 0;
+        this._onMouseDownLat = 0;
+
+        // init from camera rotation
+        this.camera.rotation.reorder('ZYX');
+        this._lon = -THREE.Math.radToDeg(this.camera.rotation.z);
+        this._lat = THREE.Math.radToDeg(this.camera.rotation.x) + 90;
+        this.updateAngles();
+
+        const domElement = view.mainLoop.gfxEngine.renderer.domElement;
+        domElement.addEventListener('mousedown', onDocumentMouseDown.bind(this), false);
+        domElement.addEventListener('mousemove', onDocumentMouseMove.bind(this), false);
+        domElement.addEventListener('mouseup', onDocumentMouseUp.bind(this), false);
+        domElement.addEventListener('keyup', onKeyUp.bind(this), true);
+        domElement.addEventListener('keydown', onKeyDown.bind(this), true);
+
+        this.view.addFrameRequester(this);
+
+        // focus policy
+        if (options.focusOnMouseOver) {
+            domElement.addEventListener('mouseover', () => domElement.focus());
+        }
+        if (options.focusOnClick) {
+            domElement.addEventListener('click', () => domElement.focus());
+        }
+    }
+
+    isUserInteracting() {
+        return this.moves.size !== 0;
+    }
+
+    updateAngles() {
+        const phi = THREE.Math.degToRad(this._lat - 90);
+        const theta = THREE.Math.degToRad(-this._lon);
+        this.camera.rotation.order = 'ZYX';
+        this.camera.rotation.x = phi;
+        this.camera.rotation.z = theta;
+
+        this.view.notifyChange(true);
+    }
+
+    update(dt, updateLoopRestarted) {
+        // if we are in a keypressed state, then update position
+
+        // dt will not be relevant when we just started rendering, we consider a 1-frame move in this case
+        if (updateLoopRestarted) {
+            dt = 16;
+        }
+
+        for (const move of this.moves) {
+            if (move.method === 'translateY') {
+                this.camera.position.z += move.sign * this.moveSpeed * dt / 1000;
+            } else {
+                this.camera[move.method](move.sign * this.moveSpeed * dt / 1000);
+            }
+        }
+
+        if (this.moves.size > 0 || this._isUserInteracting) {
```

Comment:
Sure it could have been written `if (this.moves.size || ...)` but I don't think the `> 0` harms readability.